File: data-structure/advanced/Beautiful Segments.c
#include <stdio.h>
#include <stdlib.h>

#define TYPE_EVENT 0
#define TYPE_QUERY 1

typedef long long ll;

typedef struct {
    int type;   // 0 event, 1 query
    int R;      // time dimension
    int val;    // AND value for event, K for query
    int l1;     // event: left range start; query: L
    int l2;     // event: left range end;   query: unused
    int id;     // query id, event: -1
} Item;

/* -------- Fast Input -------- */
static inline int read_int() {
    int c = getchar_unlocked();
    while (c <= ' ' && c != EOF) c = getchar_unlocked();
    int x = 0;
    while (c > ' ') {
        x = x * 10 + (c - '0');
        c = getchar_unlocked();
    }
    return x;
}

/* -------- BIT for range add + suffix sum --------
   Uses two BITs for range add / prefix sum, plus a running total so suffix_sum(L)
   = total - prefix_sum(L-1), avoiding prefix_sum(n) per query.
*/
typedef struct {
    int n;
    ll *b1;
    ll *b2;
    ll total;
} BITRange;

static inline void bit_add(ll *bit, int n, int i, ll v) {
    while (i <= n) {
        bit[i] += v;
        i += i & -i;
    }
}

static inline ll bit_sum(ll *bit, int i) {
    ll s = 0;
    while (i > 0) {
        s += bit[i];
        i -= i & -i;
    }
    return s;
}

static inline void bit_range_add(BITRange *B, int l, int r, ll v) {
    if (l > r) return;
    B->total += v * (ll)(r - l + 1);

    int n = B->n;
    bit_add(B->b1, n, l, v);
    bit_add(B->b1, n, r + 1, -v);
    bit_add(B->b2, n, l, v * (ll)(l - 1));
    bit_add(B->b2, n, r + 1, -v * (ll)r);
}

static inline ll bit_prefix_sum(BITRange *B, int i) {
    if (i <= 0) return 0;
    ll s1 = bit_sum(B->b1, i);
    ll s2 = bit_sum(B->b2, i);
    return s1 * (ll)i - s2;
}

static inline ll bit_suffix_sum(BITRange *B, int l) {
    return B->total - bit_prefix_sum(B, l - 1);
}

/* -------- Sorting by R -------- */
static int cmp_item_R(const void *pa, const void *pb) {
    const Item *a = (const Item *)pa;
    const Item *b = (const Item *)pb;
    if (a->R < b->R) return -1;
    if (a->R > b->R) return 1;
    return a->type - b->type;
}

/* -------- CDQ (merge by val) -------- */
static Item *items;
static int *ord_idx;
static int *tmp_idx;
static int *applied;
static ll *ans;
static BITRange bitR;

static void cdq(int lo, int hi) {
    if (lo >= hi) return;
    int mid = (lo + hi) >> 1;

    cdq(lo, mid);
    cdq(mid + 1, hi);

    int i = lo, j = mid + 1, k = lo;
    int app_cnt = 0;

    while (i <= mid && j <= hi) {
        int li = ord_idx[i];
        int rj = ord_idx[j];

        if (items[li].val <= items[rj].val) {
            if (items[li].type == TYPE_EVENT) {
                bit_range_add(&bitR, items[li].l1, items[li].l2, 1);
                applied[app_cnt++] = li;
            }
            tmp_idx[k++] = li;
            i++;
        } else {
            if (items[rj].type == TYPE_QUERY) {
                ans[items[rj].id] += bit_suffix_sum(&bitR, items[rj].l1);
            }
            tmp_idx[k++] = rj;
            j++;
        }
    }

    while (i <= mid) {
        int li = ord_idx[i++];
        if (items[li].type == TYPE_EVENT) {
            bit_range_add(&bitR, items[li].l1, items[li].l2, 1);
            applied[app_cnt++] = li;
        }
        tmp_idx[k++] = li;
    }

    while (j <= hi) {
        int rj = ord_idx[j++];
        if (items[rj].type == TYPE_QUERY) {
            ans[items[rj].id] += bit_suffix_sum(&bitR, items[rj].l1);
        }
        tmp_idx[k++] = rj;
    }

    // rollback
    for (int t = 0; t < app_cnt; t++) {
        int li = applied[t];
        bit_range_add(&bitR, items[li].l1, items[li].l2, -1);
    }

    // copy back
    for (int t = lo; t <= hi; t++) ord_idx[t] = tmp_idx[t];
}

int main() {
    int n = read_int();
    int q = read_int();

    int *a = (int *)malloc((n + 1) * sizeof(int));
    for (int i = 1; i <= n; i++) a[i] = read_int();

    // Max events <= 17*n (Ai < 2^17) plus q queries
    int max_items = 17 * n + q + 5;
    items = (Item *)malloc((size_t)max_items * sizeof(Item));
    int m = 0;

    // Build events per right endpoint
    int prev_val[25], prev_min[25], prev_len = 0;
    for (int j = 1; j <= n; j++) {
        int x = a[j];
        int cur_val[25], cur_min[25], cur_len = 0;

        // start with [x, j]
        cur_val[cur_len] = x;
        cur_min[cur_len] = j;
        cur_len++;

        for (int t = 0; t < prev_len; t++) {
            int nv = prev_val[t] & x;
            if (nv == cur_val[cur_len - 1]) {
                cur_min[cur_len - 1] = prev_min[t];
            } else {
                cur_val[cur_len] = nv;
                cur_min[cur_len] = prev_min[t];
                cur_len++;
            }
        }

        int prev_l = j + 1;
        for (int t = 0; t < cur_len; t++) {
            int mn = cur_min[t];
            items[m].type = TYPE_EVENT;
            items[m].R = j;
            items[m].val = cur_val[t];
            items[m].l1 = mn;
            items[m].l2 = prev_l - 1;
            items[m].id = -1;
            m++;
            prev_l = mn;
        }

        // move cur -> prev
        prev_len = cur_len;
        for (int t = 0; t < cur_len; t++) {
            prev_val[t] = cur_val[t];
            prev_min[t] = cur_min[t];
        }
    }

    // Add queries
    ans = (ll *)calloc((size_t)q, sizeof(ll));
    for (int i = 0; i < q; i++) {
        int L = read_int();
        int R = read_int();
        int K = read_int();
        items[m].type = TYPE_QUERY;
        items[m].R = R;
        items[m].val = K;
        items[m].l1 = L;
        items[m].l2 = 0;
        items[m].id = i;
        m++;
    }

    free(a);

    // Sort items by R once
    qsort(items, (size_t)m, sizeof(Item), cmp_item_R);

    // Prepare CDQ arrays
    ord_idx = (int *)malloc((size_t)m * sizeof(int));
    tmp_idx = (int *)malloc((size_t)m * sizeof(int));
    applied = (int *)malloc((size_t)m * sizeof(int));
    for (int i = 0; i < m; i++) ord_idx[i] = i;

    // BIT init
    bitR.n = n;
    bitR.b1 = (ll *)calloc((size_t)(n + 2), sizeof(ll));
    bitR.b2 = (ll *)calloc((size_t)(n + 2), sizeof(ll));
    bitR.total = 0;

    // Run CDQ
    cdq(0, m - 1);

    // Output
    for (int i = 0; i < q; i++) {
        printf("%lld\n", ans[i]);
    }

    // Cleanup
    free(items);
    free(ord_idx);
    free(tmp_idx);
    free(applied);
    free(ans);
    free(bitR.b1);
    free(bitR.b2);

    return 0;
}

File: data-structure/advanced/Beautiful Segments.py
import sys
sys.setrecursionlimit(1_000_000)

TYPE_EVENT = 0
TYPE_QUERY = 1

class BITRange:
    # Range add, prefix sum using 2 BITs, plus a running total for suffix queries.
    __slots__ = ("n", "b1", "b2", "total")
    def __init__(self, n):
        self.n = n
        self.b1 = [0] * (n + 2)
        self.b2 = [0] * (n + 2)
        self.total = 0

    def _add(self, bit, i, v):
        n = self.n
        while i <= n:
            bit[i] += v
            i += i & -i

    def range_add(self, l, r, v):
        if l > r:
            return
        self.total += v * (r - l + 1)
        b1 = self.b1
        b2 = self.b2
        n = self.n

        i = l
        while i <= n:
            b1[i] += v
            b2[i] += v * (l - 1)
            i += i & -i

        i = r + 1
        nv = -v
        while i <= n:
            b1[i] += nv
            b2[i] += nv * r
            i += i & -i

    def prefix_sum(self, i):
        if i <= 0:
            return 0
        b1 = self.b1
        b2 = self.b2
        s1 = 0
        s2 = 0
        x = i
        while x > 0:
            s1 += b1[x]
            s2 += b2[x]
            x -= x & -x
        return s1 * i - s2

    def suffix_sum(self, l):
        return self.total - self.prefix_sum(l - 1)


def solve():
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    it = iter(data)
    n = next(it)
    q = next(it)
    a = [0] * (n + 1)
    for i in range(1, n + 1):
        a[i] = next(it)

    # Parallel arrays for items to reduce overhead
    typ = []
    Rv  = []
    val = []
    l1  = []
    l2  = []
    qid = []

    # Build events: <= 17 distinct ANDs per right endpoint (Ai < 2^17)
    prev = []
    for j in range(1, n + 1):
        x = a[j]
        cur = [(x, j)]
        for v, mn in prev:
            nv = v & x
            if nv == cur[-1][0]:
                cur[-1] = (nv, mn)
            else:
                cur.append((nv, mn))

        prev_l = j + 1
        for v, mn in cur:
            typ.append(TYPE_EVENT)
            Rv.append(j)
            val.append(v)
            l1.append(mn)
            l2.append(prev_l - 1)
            qid.append(-1)
            prev_l = mn
        prev = cur

    ans = [0] * q
    for qi in range(q):
        L = next(it)
        R = next(it)
        K = next(it)
        typ.append(TYPE_QUERY)
        Rv.append(R)
        val.append(K)
        l1.append(L)   # store L here
        l2.append(0)
        qid.append(qi)

    m = len(typ)

    # Sort by R once
    idx = list(range(m))
    idx.sort(key=lambda i: Rv[i])

    typ2 = [0] * m
    R2   = [0] * m
    val2 = [0] * m
    l12  = [0] * m
    l22  = [0] * m
    qid2 = [0] * m
    for k, i in enumerate(idx):
        typ2[k] = typ[i]
        R2[k]   = Rv[i]
        val2[k] = val[i]
        l12[k]  = l1[i]
        l22[k]  = l2[i]
        qid2[k] = qid[i]
    typ, Rv, val, l1, l2, qid = typ2, R2, val2, l12, l22, qid2

    # CDQ merge sort by val on ord_idx
    ord_idx = list(range(m))
    tmp = [0] * m
    bit = BITRange(n)

    def cdq(lo, hi):
        if lo >= hi:
            return
        mid = (lo + hi) >> 1
        cdq(lo, mid)
        cdq(mid + 1, hi)

        i = lo
        j = mid + 1
        k = lo

        applied_l1 = []
        applied_l2 = []

        while i <= mid and j <= hi:
            li = ord_idx[i]
            rj = ord_idx[j]
            if val[li] <= val[rj]:
                if typ[li] == TYPE_EVENT:
                    ll = l1[li]
                    rr = l2[li]
                    bit.range_add(ll, rr, 1)
                    applied_l1.append(ll)
                    applied_l2.append(rr)
                tmp[k] = li
                i += 1
            else:
                if typ[rj] == TYPE_QUERY:
                    ans[qid[rj]] += bit.suffix_sum(l1[rj])
                tmp[k] = rj
                j += 1
            k += 1

        while i <= mid:
            li = ord_idx[i]
            if typ[li] == TYPE_EVENT:
                ll = l1[li]
                rr = l2[li]
                bit.range_add(ll, rr, 1)
                applied_l1.append(ll)
                applied_l2.append(rr)
            tmp[k] = li
            i += 1
            k += 1

        while j <= hi:
            rj = ord_idx[j]
            if typ[rj] == TYPE_QUERY:
                ans[qid[rj]] += bit.suffix_sum(l1[rj])
            tmp[k] = rj
            j += 1
            k += 1

        # rollback
        for t in range(len(applied_l1)):
            bit.range_add(applied_l1[t], applied_l2[t], -1)

        ord_idx[lo:hi + 1] = tmp[lo:hi + 1]

    cdq(0, m - 1)

    sys.stdout.write("\n".join(map(str, ans)))

if __name__ == "__main__":
    solve()


# Code times out